harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,1054 @@
+                torch DataLoader.
+            epoch_idx (integer): index of the current epoch among all the batches processed
+                per device (slot) since the start of training.
+            batch_idx (integer): index of the current batch among all the epoches processed
+                per device (slot) since the start of training.
+        Returns:
+            torch.Tensor or Dict[str, Any]:
+                training metrics to return.
+        """
+        pass
+
+    @abstractmethod
+    def build_training_data_loader(self) -> Optional[pytorch.DataLoader]:
+        """
+        Defines the data loader to use during training.
+
+        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+
+        If using data parallel training, the batch size should be per gpu batch size.
+        If using gradient aggregation, the dataloader should return batches with
+        `train_micro_batch_size_per_gpu` samples each.
+        """
+        pass
+
+    @abstractmethod
+    def build_validation_data_loader(self) -> Optional[pytorch.DataLoader]:
+        """
+        Defines the data loader to use during validation.
+
+        Must return an instance of :py:class:`determined.pytorch.DataLoader`.
+
+        If using data parallel training, the batch size should be per gpu batch size.
+        If using gradient aggregation, the dataloader should return batches with
+        `train_micro_batch_size_per_gpu` samples each.
+        """
+        pass
+
+    def build_callbacks(self) -> Dict[str, pytorch.PyTorchCallback]:
+        """
+        Defines a dictionary of string names to callbacks to be used during
+        training and/or validation.
+
+        The string name will be used as the key to save and restore callback
+        state for any callback that defines :meth:`load_state_dict` and :meth:`state_dict`.
+        """
+        return {}
+
+    def evaluate_batch(
+        self, dataloader_iter: Optional[Iterator[torch.utils.data.DataLoader]], batch_idx: int
+    ) -> Dict[str, Any]:
+        """
+        Calculate validation metrics for a batch and return them as a
+        dictionary mapping metric names to metric values. Per-batch validation metrics
+        are reduced (aggregated) to produce a single set of validation metrics for the
+        entire validation set (see :meth:`evaluation_reducer`).
+
+        The metrics returned from this function must be JSON-serializable.
+
+        Arguments:
+            batch (Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor):
+                batch of data for evaluating.
+        """
+        pass
+
+    def evaluation_reducer(self) -> Union[pytorch.Reducer, Dict[str, pytorch.Reducer]]:
+        """
+        Return a reducer for all evaluation metrics, or a dict mapping metric
+        names to individual reducers. Defaults to :obj:`determined.pytorch.Reducer.AVG`.
+        """
+        return pytorch.Reducer.AVG
```

Comment:
can we delete this method?  Is it actually useful in real life?

My impression is that users like custom reducers but I don't know that users ever use this method.

Reply:
deleted